src/dodal/devices/fast_grid_scan.py

```diff
@@ -181,181 +193,165 @@ def non_integer_dwell_time(cls, dwell_time_ms: float) -> float:
-
-    # Kickoff timeout in seconds
-    KICKOFF_TIMEOUT: float = 5.0
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        def set_expected_images(*_, **__):
-            x = int(self.x_steps.get())
-            y = int(self.y_steps.get())
-            z = int(self.z_steps.get())
-            first_grid = x * y
-            second_grid = x * z
-            self.expected_images.put(first_grid + second_grid)
-
-        self.x_steps.subscribe(set_expected_images)
-        self.y_steps.subscribe(set_expected_images)
-        self.z_steps.subscribe(set_expected_images)
-
-    def is_invalid(self) -> bool:
-        if "GONP" in self.scan_invalid.pvname:
-            return False
-        return bool(self.scan_invalid.get())
-
-    def kickoff(self) -> StatusBase:
-        st = DeviceStatus(device=self, timeout=self.KICKOFF_TIMEOUT)
-
-        def scan():
-            try:
-                curr_prog = self.motion_program.program_number.get()
-                running = self.motion_program.running.get()
-                if running:
-                    LOGGER.info(f"Motion program {curr_prog} still running, waiting...")
-                    await_value(self.motion_program.running, 0).wait()
-                LOGGER.debug("Running scan")
-                self.run_cmd.put(1)
-                LOGGER.info("Waiting for FGS to start")
-                await_value(self.status, 1).wait()
-                st.set_finished()
-                LOGGER.debug(f"{st} finished, exiting FGS kickoff thread")
-            except Exception as e:
-                st.set_exception(e)
-
-        threading.Thread(target=scan, daemon=True).start()
-        LOGGER.info("Returning FGS kickoff status")
-        return st
-
-    def complete(self) -> DeviceStatus:
-        return GridScanCompleteStatus(self)
-
-    def collect(self):
-        return {}
-
-    def describe_collect(self):
-        return {}
-
-
-def set_fast_grid_scan_params(scan: FastGridScan, params: GridScanParams):
-    yield from mv(
-        scan.x_steps,
-        params.x_steps,
-        scan.y_steps,
-        params.y_steps,
-        scan.z_steps,
-        params.z_steps,
-        scan.x_step_size,
-        params.x_step_size,
-        scan.y_step_size,
-        params.y_step_size,
-        scan.z_step_size,
-        params.z_step_size,
-        scan.dwell_time_ms,
-        params.dwell_time_ms,
-        scan.x_start,
-        params.x_start,
-        scan.y1_start,
-        params.y1_start,
-        scan.y2_start,
-        params.y2_start,
-        scan.z1_start,
-        params.z1_start,
-        scan.z2_start,
-        params.z2_start,
-        scan.position_counter,
-        0,
-    )
+def set_fast_grid_scan_params(scan: FastGridScanCommon, params: GridScanParamsCommon):
+    assert set(params.get_param_positions().keys()) == set(
+        scan.movable_params.keys()
+    ), "Scan parameters don't match the scan device"
```

Comment:
Could: We can get the type checker to check this.
```python

ParamType = TypeVar("ParamType", bound=GridScanParamsCommon)

class FastGridScanCommon(Generic[ParamType], StandardReadable, ABC):
    ...
    
class ZebraFastGridScan(FastGridScanCommon[ZebraGridScanParams]):
    ....
   
class PandAFastGridScan(FastGridScanCommon[PandAGridScanParams]):
    ....
    
def set_fast_grid_scan_params(scan: FastGridScanCommon[ParamType], params: ParamType):
   ....
```

Which will make the type checker highlight code like
```python
    assert isinstance(parameters.experiment_params, PandAGridScanParams)
    yield from set_flyscan_params(
        fgs_composite.zebra_fast_grid_scan, parameters.experiment_params
    )
```

I would be interested to see if @callumforrester or @dperl-dls have thoughts on using the type system like this.

Reply:
I like it